src/roman_datamodels/stnode/core/_d_node.py

```diff
@@ -0,0 +1,424 @@
+from __future__ import annotations
+
+import datetime
+import re
+from abc import ABC, abstractmethod
+from collections.abc import Callable, Generator, Iterator, MutableMapping
+from enum import Enum
+from inspect import getattr_static, signature
+from typing import TYPE_CHECKING, Annotated, Any, TypeVar
+
+import numpy as np
+from asdf import AsdfFile
+from asdf.lazy_nodes import AsdfDictNode, AsdfListNode
+from asdf.tags.core import ndarray
+from astropy.time import Time
+
+from ._descriptors import classproperty
+from ._field import field
+from ._mixins import AsdfNodeMixin, FlushOptions
+
+if TYPE_CHECKING:
+    from ..rad import TagMixin
+
+__all__ = [
+    "DNode",
+    "MissingFieldError",
+    "PatternDNode",
+]
+
+_T = TypeVar("_T")
+
+
+class MissingFieldError(AttributeError):
+    """Error when a field is attempted to be accessed but is missing"""
+
+
+# Once we are >3.11 -> DNode[T] can replace the Generic[T] in the class definition
+class DNode(AsdfNodeMixin[_T], MutableMapping[str, _T]):
```

Comment:
This is the `DNode` object from before. I did fully strip out the last vestiges of the automatic validation. The main changes are mostly additions so that introspection works correctly.